File: base/yarn/app/model/circuit_breaker.py
"""Per-endpoint circuit breaker for model providers.

Ported from base/planner/app/model_client.py with the same state machine:
closed (normal) -> open (all calls rejected) -> half-open (probe).
"""

from __future__ import annotations

import logging
import time

logger = logging.getLogger("yarn.model.breaker")

CLOSED = "closed"
OPEN = "open"
HALF_OPEN = "half_open"


class CircuitBreaker:
    __slots__ = (
        "_failure_count",
        "_half_open_calls",
        "_last_failure_time",
        "_state",
        "failure_threshold",
        "half_open_max",
        "name",
        "recovery_timeout",
    )

    def __init__(
        self,
        name: str,
        failure_threshold: int = 5,
        recovery_timeout: float = 60.0,
        half_open_max: int = 1,
    ) -> None:
        self.name = name
        self.failure_threshold = failure_threshold
        self.recovery_timeout = recovery_timeout
        self.half_open_max = half_open_max
        self._state = CLOSED
        self._failure_count = 0
        self._last_failure_time = 0.0
        self._half_open_calls = 0

    @property
    def state(self) -> str:
        if self._state == OPEN:
            if time.time() - self._last_failure_time >= self.recovery_timeout:
                self._state = HALF_OPEN
                self._half_open_calls = 0
                logger.info("Circuit breaker %s: OPEN -> HALF_OPEN", self.name)
        return self._state

    def allow_request(self) -> bool:
        s = self.state
        if s == CLOSED:
            return True
        if s == HALF_OPEN:
            if self._half_open_calls < self.half_open_max:
                self._half_open_calls += 1
                return True
            return False
        return False

    def record_success(self) -> None:
        if self._state in (HALF_OPEN, OPEN):
            logger.info("Circuit breaker %s: %s -> CLOSED", self.name, self._state)
        self._state = CLOSED
        self._failure_count = 0

    def record_failure(self) -> None:
        self._failure_count += 1
        self._last_failure_time = time.time()
        if self._state == HALF_OPEN:
            self._state = OPEN
            logger.warning("Circuit breaker %s: HALF_OPEN -> OPEN", self.name)
        elif self._failure_count >= self.failure_threshold:
            self._state = OPEN
            logger.warning(
                "Circuit breaker %s: CLOSED -> OPEN (failures=%d)",
                self.name,
                self._failure_count,
            )

File: base/yarn-ts/src/providers/circuit-breaker.ts
/**
 * Per-endpoint circuit breaker for model providers.
 *
 * State machine: closed (normal) → open (reject fast) → half-open (probe).
 * Ported from the retired planner runtime with identical semantics.
 * Keyed by `${modelId}:${orgId}` so one tenant's failures cannot deny
 * service to others.
 */

export const CLOSED = "closed" as const;
export const OPEN = "open" as const;
export const HALF_OPEN = "half_open" as const;

export type BreakerState = typeof CLOSED | typeof OPEN | typeof HALF_OPEN;

export interface CircuitBreakerStats {
  breakerCount: number;
  openBreakers: string[];
  halfOpenBreakers: string[];
}

class CircuitBreaker {
  private state: BreakerState = CLOSED;
  private failureCount = 0;
  private lastFailureTime = 0;
  private halfOpenCalls = 0;

  constructor(
    readonly name: string,
    private readonly failureThreshold: number,
    private readonly recoveryTimeoutMs: number,
    private readonly halfOpenMax: number,
  ) {}

  getState(): BreakerState {
    if (this.state === OPEN) {
      if (Date.now() - this.lastFailureTime >= this.recoveryTimeoutMs) {
        this.state = HALF_OPEN;
        this.halfOpenCalls = 0;
      }
    }
    return this.state;
  }

  allowRequest(): boolean {
    const s = this.getState();
    if (s === CLOSED) return true;
    if (s === HALF_OPEN) {
      if (this.halfOpenCalls < this.halfOpenMax) {
        this.halfOpenCalls += 1;
        return true;
      }
      return false;
    }
    return false;
  }

  recordSuccess(): void {
    this.state = CLOSED;
    this.failureCount = 0;
  }

  recordFailure(): void {
    this.failureCount += 1;
    this.lastFailureTime = Date.now();
    if (this.state === HALF_OPEN) {
      this.state = OPEN;
    } else if (this.failureCount >= this.failureThreshold) {
      this.state = OPEN;
    }
  }
}

const MAX_BREAKERS = 2048;

export class CircuitBreakerRegistry {
  private readonly breakers = new Map<string, CircuitBreaker>();
  private readonly failureThreshold: number;
  private readonly recoveryTimeoutMs: number;
  private readonly halfOpenMax: number;

  constructor(opts?: {
    failureThreshold?: number;
    recoveryTimeoutMs?: number;
    halfOpenMax?: number;
  }) {
    this.failureThreshold = opts?.failureThreshold ?? 5;
    this.recoveryTimeoutMs = opts?.recoveryTimeoutMs ?? 60_000;
    this.halfOpenMax = opts?.halfOpenMax ?? 1;
  }

  private getBreaker(key: string): CircuitBreaker {
    let b = this.breakers.get(key);
    if (!b) {
      if (this.breakers.size >= MAX_BREAKERS) {
        const oldest = this.breakers.keys().next().value!;
        this.breakers.delete(oldest);
      }
      b = new CircuitBreaker(
        key,
        this.failureThreshold,
        this.recoveryTimeoutMs,
        this.halfOpenMax,
      );
      this.breakers.set(key, b);
    }
    return b;
  }

  allowRequest(modelId: string, orgId: string): boolean {
    const key = orgId ? `${modelId}:${orgId}` : modelId;
    return this.getBreaker(key).allowRequest();
  }

  recordSuccess(modelId: string, orgId: string): void {
    const key = orgId ? `${modelId}:${orgId}` : modelId;
    this.getBreaker(key).recordSuccess();
  }

  recordFailure(modelId: string, orgId: string): void {
    const key = orgId ? `${modelId}:${orgId}` : modelId;
    this.getBreaker(key).recordFailure();
  }

  getStats(): CircuitBreakerStats {
    const openBreakers: string[] = [];
    const halfOpenBreakers: string[] = [];
    for (const [key, b] of this.breakers) {
      const s = b.getState();
      if (s === OPEN) openBreakers.push(key);
      else if (s === HALF_OPEN) halfOpenBreakers.push(key);
    }
    return {
      breakerCount: this.breakers.size,
      openBreakers,
      halfOpenBreakers,
    };
  }
}
